src/languages/css-source-code.js

```diff
@@ -0,0 +1,309 @@
+			) {
+				const message = `${label} comment should not span multiple lines.`;
+
+				problems.push({
+					ruleId: null,
+					message,
+					loc: comment.loc,
+				});
+				return;
+			}
+
+			switch (label) {
+				case "eslint-disable":
+				case "eslint-enable":
+				case "eslint-disable-next-line":
+				case "eslint-disable-line": {
+					const directiveType = label.slice("eslint-".length);
+
+					directives.push(
+						new Directive({
+							type: /** @type {DirectiveType} */ (directiveType),
+							node: comment,
+							value,
+							justification,
+						}),
+					);
+				}
+
+				// no default
+			}
+		});
+
+		return { problems, directives };
+	}
+
+	/**
+	 * Returns inline rule configurations along with any problems
+	 * encountered while parsing the configurations.
+	 * @returns {{problems:Array<FileProblem>,configs:Array<{config:{rules:RulesConfig},loc:SourceLocation}>}} Information
+	 *      that ESLint needs to further process the rule configurations.
+	 */
+	applyInlineConfig() {
+		const problems = [];
+		const configs = [];
+
+		this.getInlineConfigNodes().forEach(comment => {
+			const { label, value } = commentParser.parseDirective(
+				comment.value,
+			);
+
+			if (label === "eslint") {
+				const parseResult = commentParser.parseJSONLikeConfig(value);
+
+				if (parseResult.ok) {
+					configs.push({
+						config: {
+							rules: parseResult.config,
+						},
+						loc: comment.loc,
+					});
+				} else {
+					problems.push({
+						ruleId: null,
+						message:
+							/** @type {{ok: false, error: { message: string }}} */ (
+								parseResult
+							).error.message,
+						loc: comment.loc,
+					});
+				}
+			}
+		});
+
+		return {
+			configs,
+			problems,
+		};
+	}
+
+	/**
+	 * Returns the parent of the given node.
+	 * @param {CssNodePlain} node The node to get the parent of.
+	 * @returns {CssNodePlain|undefined} The parent of the node.
+	 */
+	getParent(node) {
+		return this.#parents.get(node);
+	}
+
+	/**
+	 * Traverse the source code and return the steps that were taken.
+	 * @returns {Iterable<CSSTraversalStep>} The steps that were taken while traversing the source code.
+	 */
+	traverse() {
+		// Because the AST doesn't mutate, we can cache the steps
+		if (this.#steps) {
+			return this.#steps.values();
+		}
+
+		/** @type {Array<CSSTraversalStep>} */
+		const steps = (this.#steps = []);
+
+		// Note: We can't use `walk` from `css-tree` because it uses `CssNode` instead of `CssNodePlain`
+
+		const visit = (node, parent) => {
+			// first set the parent
+			this.#parents.set(node, parent);
+
+			// then add the step
+			steps.push(
+				new CSSTraversalStep({
+					target: node,
+					phase: 1,
+					args: [node, parent],
+				}),
+			);
+
+			// then visit the children
+			for (const key of visitorKeys[node.type] || []) {
+				const child = node[key];
+
+				if (child) {
+					if (Array.isArray(child)) {
+						child.forEach(grandchild => {
+							visit(grandchild, node);
+						});
```

Comment:
Is it guaranteed that all array items are always AST nodes, not e.g., `null` as can be the case in ESTree (for example, in `ArrayExpression#elements`)?

Reply:
As far as I can tell, yes.